tree-sitter-stack-graphs/src/functions.rs

```diff
@@ -0,0 +1,194 @@
+            parameters.finish()?;
+
+            let path = path.file_name();
+            Ok(path
+                .map(|p| p.to_str().unwrap().into())
+                .unwrap_or(Value::Null))
+        }
+    }
+
+    impl Function for PathFileStem {
+        fn call(
+            &mut self,
+            _graph: &mut Graph,
+            _source: &str,
+            parameters: &mut dyn Parameters,
+        ) -> Result<Value, ExecutionError> {
+            let path = PathBuf::from(parameters.param()?.into_string()?);
+            parameters.finish()?;
+
+            let path = path.file_stem();
+            Ok(path
+                .map(|p| p.to_str().unwrap().into())
+                .unwrap_or(Value::Null))
+        }
+    }
+
+    impl Function for PathJoin {
+        fn call(
+            &mut self,
+            _graph: &mut Graph,
+            _source: &str,
+            parameters: &mut dyn Parameters,
+        ) -> Result<Value, ExecutionError> {
+            let mut path = PathBuf::new();
+            while let Ok(component) = parameters.param() {
+                path = path.join(component.into_string()?);
+            }
+
+            Ok(path.to_str().unwrap().into())
+        }
+    }
+
+    impl Function for PathNormalize {
+        fn call(
+            &mut self,
+            _graph: &mut Graph,
+            _source: &str,
+            parameters: &mut dyn Parameters,
+        ) -> Result<Value, ExecutionError> {
+            let path = PathBuf::from(parameters.param()?.into_string()?);
+            parameters.finish()?;
+
+            let path = normalize_path(&path);
+
+            Ok(path.to_str().unwrap().into())
+        }
+    }
+
+    impl Function for PathSplit {
+        fn call(
+            &mut self,
+            _graph: &mut Graph,
+            _source: &str,
+            parameters: &mut dyn Parameters,
+        ) -> Result<Value, ExecutionError> {
+            let path = PathBuf::from(parameters.param()?.into_string()?);
+            parameters.finish()?;
+
+            let components = path
+                .components()
+                .map(|c| c.as_os_str().to_str().unwrap().into())
+                .collect::<Vec<_>>();
+            Ok(components.into())
+        }
+    }
+
+    /// Normalize a path, removing things like `.` and `..`.
+    ///
+    /// CAUTION: This does not resolve symlinks (unlike
+    /// [`std::fs::canonicalize`]). This may cause incorrect or surprising
+    /// behavior at times. This should be used carefully. Unfortunately,
+    /// [`std::fs::canonicalize`] can be hard to use correctly, since it can often
+    /// fail, or on Windows returns annoying device paths. This is a problem Cargo
+    /// needs to improve on.
+    // Copied from Cargo
+    // https://github.com/rust-lang/cargo/blob/e515c3277bf0681bfc79a9e763861bfe26bb05db/crates/cargo-util/src/paths.rs#L73-L106
+    // Licensed under MIT license & Apache License (Version 2.0)
```

Comment:
I take it linking against Cargo is undesirable, and/or it's not exported?

Reply:
Not sure if it's exported, but linking against Cargo for a single functions seemed a bad idea. As a matter of fact, I found this through an online discussion where people also wondered if this shouldn't just be in the Path API 🤷🏻‍♂️